### ISSUE TITLE
Risk 6 key is missing in reported keys.

### ISSUE THREAD

**Issue by Tho-Mat** [corona-warn-app/cwa-server#723]
## Describe the bug
In the hour file 2020-08-22-hour-17.zip created at 20:05 German Time one user reports:
8 21-08; 8 20-08; 8 19-08; 5 18-08; 3 17-08; 1 16-08...1  09-08

The first key a user reports always has a risk of 6. Normally for the day before the report date.
I assume, he also reports a key for the 22-08, that was deleted (by server or app) since it was in the "future".

Possible reasons:
User has a wrong time zone or time on his device. 
So his device was already on the 23-08 as he submits his keys.
So he may also have reported wrong dates for the other keys. (+1 day)

## Questions
Is the date important for checking a key, or is only the key checked?

- If the date **is important**, then no match will be found on other devices if wrong dates are submitted.
- If the date **is not important**, then a match will be found on other devices and the only problem is, that his most resent key is not in the database even it/was is available. 

Can a user report 14 key?

Who deletes the key?

Is there a mechanism implemented, to check, that the sending device has the right time/date?

## Comments
In an extrem case the user has a date more than 14 day in the future or the past. 
In this case no keys will be included in the database. 
I do not know how this can be fixed, if the OS uses the device time and not the correct time.
The OS could get the correct time via internet or GPS, so that should not happen.

[2020-08-22-hour-17.zip](https://github.com/corona-warn-app/cwa-server/files/5113615/2020-08-22-hour-17.zip)

**Comment by michael-burwig** [comment]
Hi @Tho-Mat, thanks for reporting this here and sorry for the delayed response. After talking to some folks, the root cause that you describe could indeed be the culprit here. But I am still investigating this a little further.

Just to quickly address some of the questions you raise:
- Yes, users can (and as far as I know also do) submit 14 keys at a time.
- If an incorrect end device time were to be at fault, I suspect that the server would delete the key in question, if it fails the respective sanity check. (I still have to verify this)
- We do not check whether an end device has its time set correctly.

As time permits it I will look into the matter more deeply.

**Comment by daimpi** [comment]
> - Yes, users can (and as far as I know also do) submit 14 keys at a time.

Are you sure about this? My understanding is that CWA only uploads 13 keys at a maximum. The TEK from today will simply not be uploaded, but rather only those up to 13 TEKs from the days before. This "last day upload" seems like something the devs would like to implement at some point, but I don't think this is live currently: https://github.com/corona-warn-app/cwa-app-android/issues/678#issuecomment-650760871.

This is also what you find e.g. [here](https://ctt.pfstr.de/) for the number of uploaded keys per user. The maximal number is 13.

**Comment by michael-burwig** [comment]
I remember that it used to be 13 (for the initial submission) but understood from conversations that I had today that it changed at some point. ...I might have misunderstood this part. Will check again tomorrow.

**Comment by michael-burwig** [comment]
Following up on @daimpi's remark:
While mobile devices (at least Android) upload up to 14 keys, the server checks whether the generation timestamp exceeds the retention period and discards the respective key(s). In practice this means that the server stores and distributes a maximum of 13 keys per user.

As to the initially reported issue of the missing key, it is still on my list to investigate this further.

**Comment by daimpi** [comment]
Thanks for the update 🙂 
Just to clarify:
> While mobile devices (at least Android) upload up to 14 keys

this doesn't include today's key, but only those from the 14 days prior, right?

**Comment by michael-burwig** [comment]
> this doesn't include today's key, but only those from the 14 days prior, right?

Correct, whether this makes sense or not might be another topic 😉 .

**Comment by Tho-Mat** [comment]
@michael-burwig 
any news about the issue?

**Comment by michael-burwig** [comment]
I could not find anything suspicious on the backend side with regards to mishandling of time zones.

As @Tho-Mat pointed out in his initial post, an incorrectly set device time (not time zone) could be at fault for dropping the youngest key in this case.

If this problem is addressed in the future, I assume that it will be driven by the app team. We will follow up with them but at this point in time, it appears to me that the server components are doing what they are supposed to do.

Hence I am suggesting to close #743 and focus the discussion in [cwa-app-ios#1097](https://github.com/corona-warn-app/cwa-app-ios/issues/1097).

**Comment by Tho-Mat** [comment]
@michael-burwig 
So can you confirm that the keys where send without the 6?

**Comment by EvgeniiSkrebtcov** [comment]
@Tho-Mat,
As submissions are not logged for data privacy reasons, we do not have any means of checking, other than looking at the output files. Therefore, proposed explanation appears valid.

**Comment by Tho-Mat** [comment]
@EvgeniiSkrebtcov 
Thanks for answer. 
So for this case it's not possible the check.

I understand the importance of privacy, but
I do not see what additional data protection provisions would be violated, if the deletion of keys would be logged. 
The keys taken over, their risk values per date and the hour of the upload are known. 
 
It would be enough to log only the hour and the process of deletion. The keys do not need to be saved. 
So e.g. 
On 2020-07-04 in the period from 14:00-15:00 
five keys with risk value 6 from 2020-07-03 were not taken over. 
Reason: already in database / Date to high / Date to low / unknown.

This would have the great advantage of better analyzing certain constellations that can lead to errors.

**Comment by EvgeniiSkrebtcov** [comment]
Hello @Tho-Mat ,

Thank you for the proposal. We logged it here https://github.com/corona-warn-app/cwa-server/issues/764 and will review on a next team meeting.

**Comment by EvgeniiSkrebtcov** [comment]
Hello @Tho-Mat,

Together with the mobile team, we performed an additional testing round targeted to reproduce this behavior on out test systems. As a result, we can now confirm that submissions from the mobile devices with the system time set to be 1 day in the future results in the exact data pattern you observed in productive system:

- Transmission Risk Level (TRL) 6 key has a rolling_start_interval_number as 'today' midnight;
- One TRL 8 key has a rolling_start_interval_number as 'yesterday';

As I explained in https://github.com/corona-warn-app/cwa-server/issues/755#issuecomment-692942406, this results in a situation where server embargoes TRL 6 keys, until the key expires +2h, while remaining keys are distributed with the next distribution run. In all affected hourly packages we observed TRL 8 keys with the rolling_start_interval_number = 'yesterday midnight' and no matching TRL 6 keys.

Embargoed TRL 6 keys are eventually included into the earliest possible distribution according to the key expiry and shifting policy.

Although the exact reason for growing number of such cases is unknown, we will proceed with the enhancement to overcome this behaviour. For further updates kindly follow https://github.com/corona-warn-app/cwa-app-ios/issues/1097

Thank you very much for your support!
